src/snowflake/snowpark/async_job.py

```diff
@@ -0,0 +1,226 @@
+#
+# Copyright (c) 2012-2022 Snowflake Computing Inc. All rights reserved.
+#
+from enum import Enum
+from typing import Iterator, List, Union
+
+import snowflake.snowpark
+from snowflake.connector.options import pandas
+from snowflake.snowpark._internal.utils import (
+    check_is_pandas_dataframe_in_to_pandas,
+    result_set_to_iter,
+    result_set_to_rows,
+)
+from snowflake.snowpark.row import Row
+
+
+class _AsyncDataType(Enum):
+    PANDAS = "pandas"
+    ROW = "row"
+    ITERATOR = "iterator"
+    PANDAS_BATCH = "pandas_batch"
+    COUNT = "count"
+    NONE_TYPE = "no_type"
+    UPDATE = "update"
+    DELETE = "delete"
+    MERGE = "merge"
+
+
+class AsyncJob:
+    """Provides a way to track an asynchronous query in Snowflake. A :class:`DataFrame` object can be evaluated asynchronously and an :class:`AsyncJob` object will be returned. With this instance, you can do:
+    - retrieve results;
+    - check the query status (still running or done);
+    - cancel the running query;
+    - retrieve the query ID and perform other operations on this query ID manually.
+    AsyncJob is strongly related to dataframe functions, to use AsyncJob, you need to create a dataframe first
```

Comment:
```suggestion
    :class:`AsyncJob` is created by action methods in :class:`DataFrame`. Therefore, to use it, you need to create a dataframe first. Here we demonstrate how to do that: 
```
combine with the line below